feat(bulk-import): integrate bulk import apis in the bulk import frontend plugin

debsmita1 opened the pull request:
**Resolves:**
https://issues.redhat.com/browse/RHIDP-2737

**GIF/Screenshots:**

Page-wise fetching and selecting rows:

https://github.com/user-attachments/assets/5c43bf76-ebb5-426d-9226-5e94fc043d15

Failed PR creation:

https://github.com/user-attachments/assets/8b0cd038-ec8b-4070-9a80-8356174858f7


Successful PR creation:

https://github.com/user-attachments/assets/b809c2c5-9e7f-43dc-98c3-45e25161f662


**Test coverage:**

<img width="967" alt="Screenshot 2024-08-13 at 5 56 36 PM" src="https://github.com/user-attachments/assets/eaa2295c-c9b6-431f-870a-49ef242d7acd">

ciiay commented:
Hi @debsmita1 , not sure if you are done with the logic part or not, but this is what I found so far. Will continue with the code review.
1. Total amount of added repositories disappears from time to time
2. When I try to add catalog-info.yaml to some repo which already contains a catalog-info.yaml file, it says "Finished and Ingested" in the Status column. This is probably the expected behavior but if the user updated content in catalog-info.yaml they may want to replace the existing one with the new one edited in the sidebar? Or we can fetch the existing catalog-info.yaml when open the Preview file sidebar?

![image](https://github.com/user-attachments/assets/7ae4ae41-0718-44ef-b708-d748c5e01eb3)

debsmita1 commented:
> Hi @debsmita1 , not sure if you are done with the logic part or not, but this is what I found so far. Will continue with the code review.
> 
> 1. Total amount of added repositories disappears from time to time

I have updated the PR, could you PTAL again

> 2. When I try to add catalog-info.yaml to some repo which already contains a catalog-info.yaml file, it says "Finished and Ingested" in the Status column. This is probably the expected behavior but if the user updated content in catalog-info.yaml they may want to replace the existing one with the new one edited in the sidebar? Or we can fetch the existing catalog-info.yaml when open the Preview file sidebar?
> 

Yes, you are right. If the component already has the catalog-info.yaml, the create import job API doesn't create a PR but simply registers it in the catalog
cc @rm3l

ciiay commented:
Hi @debsmita1 , I have verified that the total amount of Added repositories flashing issue got resolved 👍 

I merged one of the PRs raised by bulk-import plugin, and the row disappeared from the "Added repositories" table. Is this expected? Also the one with "Finished and Ingested" from last test also disappeared. Both of the disappeared repositories displayed in the "Add repositories" page as unchecked which looks a bit weird.

debsmita1 commented:
> Hi @debsmita1 , I have verified that the total amount of Added repositories flashing issue got resolved 👍
> 
> I merged one of the PRs raised by bulk-import plugin, and the row disappeared from the "Added repositories" table. Is this expected? Also the one with "Finished and Ingested" from last test also disappeared. Both of the disappeared repositories displayed in the "Add repositories" page as unchecked which looks a bit weird.

Looks like an issue with the API, after merging the PR

The API response for `http://localhost:7007/api/bulk-import-backend/import/by-repo?repo=https://github.com/debsmita1/api&defaultBranch=master` was
```
{
    "id": "https://github.com/debsmita1/api",
    "repository": {
        "url": "https://github.com/debsmita1/api",
        "name": "api",
        "organization": "debsmita1"
    },
    "approvalTool": "GIT",
    "status": null,
    "lastUpdate": "2024-08-09T05:58:26Z"
} 
```
Also, I didn't see it getting added to the catalog page

cc @rm3l

debsmita1 commented:
> Hi @debsmita1 , I have verified that the total amount of Added repositories flashing issue got resolved 👍
> 
> I merged one of the PRs raised by bulk-import plugin, and the row disappeared from the "Added repositories" table. Is this expected? Also the one with "Finished and Ingested" from last test also disappeared. Both of the disappeared repositories displayed in the "Add repositories" page as unchecked which looks a bit weird.

Looks like an issue with the API, 
After merging the PR, the API response for `http://localhost:7007/api/bulk-import-backend/import/by-repo?repo=https://github.com/debsmita1/api&defaultBranch=master` was
```
{
    "id": "https://github.com/debsmita1/api",
    "repository": {
        "url": "https://github.com/debsmita1/api",
        "name": "api",
        "organization": "debsmita1"
    },
    "approvalTool": "GIT",
    "status": null,
    "lastUpdate": "2024-08-09T05:58:26Z"
} 
```
Also, I didn't see it getting added to the catalog page

cc @rm3l

ShiranHi commented:
Hey @debsmita1 looks good to me. Some questions:

1. How do you feel about using a spinner instead of the progress bar while keeping the table header row? It might help reduce the jumping, same behavior as here: 
<img width="1130" alt="image 01" src="https://github.com/user-attachments/assets/b56393b9-26e6-4e37-a20e-93a3edef92a5">

2. I think that we can remove the "Repositories" tab appears below the header, we have only one bulk import type at this moment. What do you think?

3. Can we add a spinner of the button (to the left side of the button label) to indicate the waiting time? I think we can remove the tooltip on hover
<img width="1130" alt="image 02" src="https://github.com/user-attachments/assets/d2711f99-63a4-459d-8f8a-8e509b1ba948">

4. If users selected just one repository (using the "Organization" view), can we remove the tab component in this case?
<img width="1128" alt="image 03" src="https://github.com/user-attachments/assets/883bc3d8-c3da-4a42-99a7-1debc42c4cb6">

debsmita1 commented:
@ShiranHi Thank you for reviewing this PR. Please find my replies below the questions

> Hey @debsmita1 looks good to me. Some questions:
> 
> 1. For the "Selected repositories" table: How do you feel about using a spinner instead of the progress bar while keeping the table header row? It might help reduce the jumping, same behavior as here:
> 
> <img alt="image 01" width="1130" src="https://private-user-images.githubusercontent.com/77277517/356821414-b56393b9-26e6-4e37-a20e-93a3edef92a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NjE0NDMsIm5iZiI6MTcyMzQ2MTE0MywicGF0aCI6Ii83NzI3NzUxNy8zNTY4MjE0MTQtYjU2MzkzYjktMjZlNi00ZTM3LWEyMGUtOTNhM2VkZWY5MmE1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDExMTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTA3NDVmYjkwMWRjOTQ5Y2ZlZWRiODg1YWYwZGM2YjM3ZWUyZmZjYmMwMzI5OTQwMGU1YWU3YTQyNzU0YTVhZTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.LJp0WL_qhaO5gZNIargf2FcMZ7Ds4WPbunwoT-mRZPE">


Done
![Screenshot 2024-08-12 at 4 48 31 PM](https://github.com/user-attachments/assets/2604645d-da53-48f5-9c64-d5a68e04de3c)




> 2. I think that we can remove the "Repositories" tab appears below the header, we have only one bulk import type at this moment. What do you think?


The Figma design (https://www.figma.com/design/09Iltd57RiuqOtKWkjy8MH/Bulk-import?node-id=3-534&t=KKaedm2zO0xDtsEG-0) indicates that two additional tabs will be added in the future.




> 3. Can we add a spinner of the button (to the left side of the button label, like [here](https://www.patternfly.org/components/button/#progress-indicators)) to indicate the waiting time? I think we can remove the tooltip on hover
> 
> <img alt="image 02" width="1130" src="https://private-user-images.githubusercontent.com/77277517/356821488-d2711f99-63a4-459d-8f8a-8e509b1ba948.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NjE0NDMsIm5iZiI6MTcyMzQ2MTE0MywicGF0aCI6Ii83NzI3NzUxNy8zNTY4MjE0ODgtZDI3MTFmOTktNjNhNC00NTlkLThmOGEtOGU1MDliMWJhOTQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDExMTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFiMDU5NTBjOTdjN2MxNDBjYTJmNTczY2UzODg0NDlkNDA1OTBhNzgwMzNhMzVhMmZhMmNjMjY1ZDNlYTgwMmEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rscaKI4nxJwAR2oMjlACxz1JOGmN2RTn9LAiUd1gE7U">

Removed the tooltip and added `CircularProgress` MUI component (https://mui.com/material-ui/react-progress/#circular-color)
![Screenshot 2024-08-12 at 4 53 49 PM](https://github.com/user-attachments/assets/ee79fffa-5c24-46d0-aa48-264f564524f4)




> 4. If users selected just one repository (using the "Organization" view), can we remove the tab component in this case?
> 
> <img alt="image 03" width="1128" src="https://private-user-images.githubusercontent.com/77277517/356821531-883bc3d8-c3da-4a42-99a7-1debc42c4cb6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NjE0NDMsIm5iZiI6MTcyMzQ2MTE0MywicGF0aCI6Ii83NzI3NzUxNy8zNTY4MjE1MzEtODgzYmMzZDgtYzNkYS00YTQyLTk5YTctMWRlYmM0MmM0Y2I2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDExMTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNjNjliODYyNzcxYThiNGRiZTVjZDJlNTNkZGZhNGIxZDdkYjA3M2QwZjNjNTlkY2Q1MzI3N2UzYjQxOTVjNDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.xGsRnsry5YedMvn_9P8hDMBOPXAZ70JbCKY8Se6oKHw">

Done
![Screenshot 2024-08-12 at 4 54 59 PM](https://github.com/user-attachments/assets/5805cc80-181c-472a-b4c3-0d63d0095abe)



> 5. Let's try to prevent from this situation when we have many progress bar at the same time. Can we skeleton instead of the progress bar? or loading the whole table content when it is ready? What do you think?
> 
> <img alt="image 04" width="1130" src="https://private-user-images.githubusercontent.com/77277517/356822069-04070fa1-ab63-4226-8a9e-6855f42da3ff.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NjE0NDMsIm5iZiI6MTcyMzQ2MTE0MywicGF0aCI6Ii83NzI3NzUxNy8zNTY4MjIwNjktMDQwNzBmYTEtYWI2My00MjI2LThhOWUtNjg1NWY0MmRhM2ZmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDExMTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2YjIzODkyNzcyN2ZiZjE4YWE1NTYwYzE3NzZlYjkyM2RkMjdlYjcxNWQ4NGUyMDBiZmI2Y2MzNmJmMzY0ZTMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.3pIkeUopHotKDpCZqbWHmwfjD9bx0bOX9QpKzNi8yG0">

I am using the progress bar because it makes a separate API call to fetch the status of each repository. We do something similar in the Quay plugin. If we still want to show a skeleton till all the data is loaded then I would like to address this In a separate ticket. 


> 6. Also here, let's show the table header and use spinner/skeleton. The text on the search field should be just "Search"
> 
> <img alt="image 05" width="1130" src="https://private-user-images.githubusercontent.com/77277517/356822106-bd461acf-7287-4bf6-9c96-40ee93ed29db.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NjE0NDMsIm5iZiI6MTcyMzQ2MTE0MywicGF0aCI6Ii83NzI3NzUxNy8zNTY4MjIxMDYtYmQ0NjFhY2YtNzI4Ny00YmY2LTljOTYtNDBlZTkzZWQyOWRiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDExMTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWIzNDk4NDM1MzhmNDRlMWJhNjBmOWM3NGUyYTg4MmZjOWMzNWE3ZTljYzZiZTk1MjQ4YzJkYzIyZTJmOWQwOTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.CfRvOTP7M71YTlQTehSshZg9FTRk5CG4Qdl8DahtYSk">


Done
![Screenshot 2024-08-12 at 4 57 27 PM](https://github.com/user-attachments/assets/3dd0b7ec-7372-4b56-9592-74d9a6ab0468)




> 7. This error is great, let's just add a bit of padding between the tabs and the message ([Figma](https://www.figma.com/design/09Iltd57RiuqOtKWkjy8MH/Bulk-import?node-id=925-1030&t=rn3zQGAZgaWqtlST-1))
> 
> <img alt="image 06" width="1129" src="https://private-user-images.githubusercontent.com/77277517/356822155-dee3f2d0-7b40-490a-8fb5-8ce90904035f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM0NjE0NDMsIm5iZiI6MTcyMzQ2MTE0MywicGF0aCI6Ii83NzI3NzUxNy8zNTY4MjIxNTUtZGVlM2YyZDAtN2I0MC00OTBhLThmYjUtOGNlOTA5MDQwMzVmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODEyVDExMTIyM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTI3YjQ1NTAxYzUxMjFkOWZjODUwN2Q4OTBkZGZkNGEyMzk2NzNjMDg1ZDIyMzZlMTExODg4YThmNGI0MDVlYzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.-5FMufUz9EpmsEXyC8sf-5PtJqlpKRgU5jIOwHHsjeU">

Done
![Screenshot 2024-08-12 at 4 58 13 PM](https://github.com/user-attachments/assets/d7cf0be5-64da-4e96-ad1c-3b017b5adbe8)

ShiranHi commented:
> > 2. I think that we can remove the "Repositories" tab appears below the header, we have only one bulk import type at this moment. What do you think?
> 
> The Figma design (https://www.figma.com/design/09Iltd57RiuqOtKWkjy8MH/Bulk-import?node-id=3-534&t=KKaedm2zO0xDtsEG-0) indicates that two additional tabs will be added in the future.

It might be more effective to add the tabs after we have content for the other sections. Displaying just one tab isn't ideal. Apologies for any confusion caused in the Figma file.


> > 5. Let's try to prevent from this situation when we have many progress bar at the same time. Can we skeleton instead of the progress bar? or loading the whole table content when it is ready? What do you think?
> 
> I am using the progress bar because it makes a separate API call to fetch the status of each repository. We do something similar in the Quay plugin. If we still want to show a skeleton till all the data is loaded then I would like to address this In a separate ticket.

Alright, let's address it in a separate ticket. I think it's worth it since the current experience feels overwhelming.

invincibleJai commented on plugins/bulk-import/src/api/BulkImportBackendClient.ts:
```diff
@@ -0,0 +1,192 @@
+      `${backendUrl}/api/bulk-import-backend/repositories?pagePerIntegration=${page}&sizePerIntegration=${size}`,
+      {
+        headers: {
+          // ...(idToken && { Authorization: `Bearer ${idToken}` }),
```
If this is it not needed in the header, then we can get rid of the comment

debsmita1 replied:
my [PR](https://github.com/janus-idp/backstage-plugins/pull/2034) that enables permission check in the bulk import frontend is ready, so removing the commented lines from here

invincibleJai commented on plugins/bulk-import/src/api/BulkImportBackendClient.ts:
```diff
@@ -0,0 +1,192 @@
+    // const { token: idToken } = await this.identityApi.getCredentials();
+
+    const backendUrl = this.configApi.getString('backend.baseUrl');
+    const jsonResponse = await fetch(
+      `${backendUrl}/api/bulk-import-backend/organizations/${orgName}/repositories?pagePerIntegration=${page}&sizePerIntegration=${size}`,
+      {
+        headers: {
+          // ...(idToken && { Authorization: `Bearer ${idToken}` }),
+          'Content-Type': 'application/json',
+        },
+      },
+    );
+    if (jsonResponse.status !== 200 && jsonResponse.status !== 204) {
+      return [];
+    }
+    return jsonResponse.json();
+  }
```
This code block seems repeated , can be abstracted out into common fetcher

debsmita1 replied:
created a common dataFetcher

invincibleJai commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesForm.tsx:
```diff
@@ -1,16 +1,18 @@
-import { useFormikContext } from 'formik';
 
-import { AddRepositoriesFormValues } from '../../types';
+// import HelpIcon from '@mui/icons-material/HelpOutline';
+// import FormControlLabel from '@mui/material/FormControlLabel';
+// import Radio from '@mui/material/Radio';
+// import RadioGroup from '@mui/material/RadioGroup';
+// import Tooltip from '@mui/material/Tooltip';
+// import Typography from '@mui/material/Typography';
+// import { useFormikContext } from 'formik';
+// import { AddRepositoriesFormValues } from '../../types';
```
remove commented code block

debsmita1 replied:
I deliberately didn't remove this code. We aren't supporting ServiceNow approval in this release, but we plan to in the next one. To save extra effort later, I commented it out instead of removing it entirely.

invincibleJai commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx:
```diff
@@ -1,31 +1,137 @@
+    const importRepositories = Object.values(values.repositories).reduce(
+      (acc: CreateImportJobRepository[], repo) => {
+        acc.push({
+          approvalTool: values.approvalTool.toLocaleUpperCase(),
+          catalogEntityName:
+            repo.catalogInfoYaml?.prTemplate?.componentName ||
+            repo?.repoName ||
+            'my-component',
+          repository: {
+            url: repo.repoUrl || '',
+            name: repo.repoName || '',
+            organization: repo.orgName || '',
+            defaultBranch: repo.defaultBranch || '',
+          },
+          catalogInfoContent: yaml.stringify(
+            repo.catalogInfoYaml?.prTemplate?.yaml,
+            null,
+            2,
+          ),
+          github: {
+            pullRequest: {
+              title:
+                repo.catalogInfoYaml?.prTemplate?.prTitle ||
+                'Add catalog-info.yaml config file',
+              body:
+                repo.catalogInfoYaml?.prTemplate?.prDescription ||
+                'This pull request adds a **Backstage entity metadata file**\nto this repository so that the component can\nbe added to the [software catalog](http://localhost:3000).\nAfter this pull request is merged, the component will become available.\nFor more information, read an [overview of the Backstage software catalog](https://backstage.io/docs/features/software-catalog/).',
+            },
+          },
```
this function block is too long, can we abstract this?

invincibleJai commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesPage.tsx:
```diff
@@ -1,31 +1,137 @@
+    );
+    bulkImportApi
+      .createImportJobs(importRepositories, true)
+      .then(async (dryrunResponse: ImportJobResponse[]) => {
```
aync await over then block?

invincibleJai commented on plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx:
```diff
@@ -1,10 +1,10 @@
 import * as React from 'react';
-import { useEffect, useMemo } from 'react';
-import { useAsync } from 'react-use';
+import { useEffect } from 'react';
```
nit: good to have one import from 'react'

invincibleJai commented on plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx:
```diff
@@ -401,104 +395,130 @@ export const RepositoriesTable = ({
     return 'repositories-table';
   };
 
+  const tableBodyContent = () => {
```
this file has become too long, can we move this out?

invincibleJai commented:
Thanks @debsmita1, in dev mode I see one issue 

Click on status/refresh breaks on dev mode

<img width="1653" alt="image" src="https://github.com/user-attachments/assets/6d3ebea6-6133-4853-9ace-dace5656058d">

<img width="1094" alt="image" src="https://github.com/user-attachments/assets/e9e2eb6e-076d-4ca3-bdb2-951558564868">

debsmita1 commented:
> Thanks @debsmita1, in dev mode I see one issue
> 
> Click on status/refresh breaks on dev mode
> 
> <img alt="image" width="1653" src="https://private-user-images.githubusercontent.com/5129024/357432951-6d3ebea6-6133-4853-9ace-dace5656058d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NTcyMTQsIm5iZiI6MTcyMzU1NjkxNCwicGF0aCI6Ii81MTI5MDI0LzM1NzQzMjk1MS02ZDNlYmVhNi02MTMzLTQ4NTMtOWFjZS1kYWNlNTY1NjA1OGQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTNUMTM0ODM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTMwZWU2YTE5MDUyY2FjNTRhMWY0MDYxNmVjZjliYWJjNDcyYTlmYjEyNWY4MmFiMzlkNTNmZWZjMGUzMjVlNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Q8bruL-5DDkIeHWGKSt4ipYxHGxS81FS7F8c9VpQleM"> <img alt="image" width="1094" src="https://private-user-images.githubusercontent.com/5129024/357433253-e9e2eb6e-076d-4ca3-bdb2-951558564868.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjM1NTcyMTQsIm5iZiI6MTcyMzU1NjkxNCwicGF0aCI6Ii81MTI5MDI0LzM1NzQzMzI1My1lOWUyZWI2ZS0wNzZkLTRjYTMtYmRiMi05NTE1NTg1NjQ4NjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDgxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA4MTNUMTM0ODM0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTM0N2Q3N2Q1ZDE3ZjdkN2QzOTEzNTA0MTI0ZWFhYzYzYmRiMjU1OTU3MGJjODU0MjJjNDQwYjRlY2VlMGUxOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ESTDjgKyfOmezyT8XTnFRJ81bYDcwIxi2B6WFxhz5ms">

Fixed it! PTAL

jerolimov commented:
I first take a look from the user perspective before I will review the code as well.

Just some small ideas/feedback, that we can also fix in a follow-up -- **_if_** @ShiranHi agrees!

1. README. I was required to add this into my Root.tsx:

```
<SidebarItem ... />
<SidebarItem ... />
<BulkImportIcon />
```

Currently the `BulkImportIcon` is actual a SidebarItem. Please export just the icon as `BulkImportIcon` and if you want simplify the SidebarItem please export a `BulkImportSidebarItem`. 

2. On the "Bulk import" start page is a repository table with a "Repo URL" column.

![image](https://github.com/user-attachments/assets/6d7e7856-a548-4ea2-bfc0-4f1db200cae5)

The link should have an "external page" icon, similar to the "Selected repositories" table > URL column.

![image](https://github.com/user-attachments/assets/ce2e8a66-e83e-4536-b5e9-3f79b1c645e9)

3. When editing a Bulk Import this link is also shown and should have also such an icon.

![image](https://github.com/user-attachments/assets/4a941341-c971-4a03-a10c-045086cce655)

4. On the "Add repositories" the search field uses "Search" and placeholder, while the catalog and import list uses "Filter".

![image](https://github.com/user-attachments/assets/842b1e9c-9314-4e9d-99dd-2ee509bc05c6)

vs

![image](https://github.com/user-attachments/assets/76af575c-fcf3-4e9e-bb41-e361e39dccc6)

5. @ShiranHi @debsmita1, when adding a repository and preparing a PR, the edit button shows a complete PR preview, not just a file.

Should we rename the column to "Pull request" and the link to "Preview PR" ???

![image](https://github.com/user-attachments/assets/344a7c0b-d8e4-40e2-95a3-596f163f270b)

6. Didn't the first version only supports GitHub PRs? Should we disable the ServiceNow option for now?

![image](https://github.com/user-attachments/assets/0a9449e1-f56f-4ec7-91f3-e94f0c4dff5b)

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/RepositoriesTableBody.tsx:
```diff
@@ -0,0 +1,122 @@
+    );
+  } else if (rows?.length > 0) {
+    return (
+      <TableBody>
+        {rows.map(row => {
```
Is there a reason why this doesn't use the Table component from `@backstage/core-component`?

jerolimov replied:
Understand that we use MUI table here to customize the headers and custom selection checkboxes.

Unsure if there are alternatives with backstage-components here, but that's fine for now.

jerolimov commented on plugins/bulk-import/src/components/BulkImportIcon.tsx:
```diff
@@ -15,8 +15,6 @@ export const icon = () => {
 
 export const BulkImportIcon = () => {
-  // permission logic
-
   return (
     <SidebarItem text="Bulk import" to="bulk-import/repositories" icon={icon} />
   );
```
As mentioned in another comment. BulkImportIcon should return an Icon or should have another name.

The name confused me. If you fix this, please also update the README. :)

I'm fine if you want to open another PR for that.

jerolimov commented on plugins/bulk-import/src/components/PreviewFile/KeyValueTextField.tsx:
```diff
@@ -2,7 +2,7 @@ import React, { useState } from 'react';
 import { FormHelperText, TextField } from '@material-ui/core';
 
-import { PullRequestPreview, PullRequestPreviewData } from '../../types';
+import { PullRequestPreview, PullRequestPreviewData } from '../../types/types';
```
types/types :question:

jerolimov replied:
resolved

jerolimov reviewed (changes requested):
Hi @debsmita1, also if I found some small nit-picky things... Great PR and kudos for all these tests!!!

jerolimov commented:
It takes some more time to actual test the latest version. :shrug:  

And with this I found some more issues. As additional to https://github.com/janus-idp/backstage-plugins/pull/2013#issuecomment-2291149128 and the code review above.

7. I got an error from the backend that the auth header is missing. I could fix this locally with this change:

   ![image](https://github.com/user-attachments/assets/32ae806c-e28a-4d36-98de-c6f11d2b7918)

   It happen without and with permissions enabled, with the guest-provider and a GitHub user.

8. I also noticed also that the import and repositories list doesn't show any error if the API call fails.

   I noticed this just by checking the network traffic.

9. For me the repository list shows 1181 entries. The search didn't worked.

   Imho: I got that the API returns just one page. But then the search needs to trigger a new API call with a filter.

   If this isn't possible today we should hide the search. A search that helps my to filter just the current page isn't useful

10. Selecting all selected just the current page. (5 of 1181 entries in my case!)

    That is a common issue with pagination. Also with the biggest page size (currently 15)  I must page thought 79 pages!

    As quick solution I like to recommend to add bigger page sizes like 5, 10, 25, 50, 100.

    In the long term it would be nice if the select all option will notice the user that there are more pages and ask if the user wants to select everything. (Gmail does this very well.)

ShiranHi commented:
> 3\. When editing a Bulk Import this link is also shown and should have also such an icon.

That's right, thank you @jerolimov, let's add the icon every time we open the link in a new tab. 

> 4\.  On the "Add repositories" the search field uses "Search" and placeholder, while the catalog and import list uses "Filter".
Later I noticed that the search only works on the current page, which is IMO a reason to hide that filter at all. See below.

I think all search fields should include a "Search" placeholder and a magnifier icon. Using "Search" instead of "Filter" aligns better with user expectations and is a standard UI convention. Also making the function clear and intuitive, when talking about filters users expects to see categories/options. It also maintains consistency across different Res Hat products ([PF component](https://www.patternfly.org/components/search-input#examples)), which improving the overall user experience. 

> 5\. @ShiranHi @debsmita1, when adding a repository and preparing a PR, the edit button shows a complete PR preview, not just a file.
Should we rename the column to "Pull request" and the link to "Preview PR"

This is a good point, but since the PR hasn't been created yet, making this change might be confusing for users. I recommend keeping it as is for now and adjusting it later based on customer feedback.

> 6\. Didn't the first version only supports GitHub PRs? Should we rename Git to GitHub and disable the ServiceNow option for now?

We’ve decided to temporarily hide the entire section from the UI until we support ServiceNow. Displaying it without support would be more confusing than helpful.

debsmita1 commented on plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx:
```diff
@@ -69,83 +40,83 @@ export const RepositoriesTable = ({
-        r => r.id,
-      );
-      setSelected(selectedOrgRepoIds);
+      setSelected(values.repositories);
```
This component is being used by the Repositories Table and the RepositoriesTable on the drawer, 

For `RepositoriesTable on the drawer` , we are using the local state. Only when the user clicks on `Save` on the drawer we update the Formik state, else we don't

jerolimov replied:
As we discussed, we should try to use formik in the drawer as well, maybe with a 2nd formik context.

But I'm accepting that we don't use formik there at the moment.

debsmita1 commented on plugins/bulk-import/src/components/AddRepositories/CatalogInfoStatus.tsx:
```diff
@@ -1,45 +1,95 @@
+    return (
+      <StatusRunning>
+        <span
+          style={{ fontWeight: '400', fontSize: '0.875rem', color: '#181818' }}
```
Yup

![Screenshot 2024-08-19 at 11 47 48 AM](https://github.com/user-attachments/assets/9d138218-d030-4907-a6e9-f9d01124a088)

jerolimov replied:
mark it as resolved to merge this PR, @debsmita1 will check if we can address this in a follow up PR.

debsmita1 commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesDrawer.tsx:
```diff
@@ -60,24 +56,37 @@ export const AddRepositoriesDrawer = ({
     onClose();
   };
 
+  React.useEffect(() => {
+    const sr = Object.values(values.repositories).reduce(
+      (acc, repo) =>
+        repo.orgName === orgData?.orgName ? { ...acc, [repo.id]: repo } : acc,
+      {},
+    );
+    setSelectedRepos(sr);
+  }, [orgData?.orgName, values.repositories]);
```
Yes

jerolimov replied:
We discussed this and it's not mandatory to change this.

But I checked the code now again and it looks like this tries to change the initial state, or?

debsmita1 replied:
yes, when the drawer component mounts, it reads the selected repositories from the formik context and selects the rows on the drawer

jerolimov replied:
If this is just for the first time, then you should initialize this data when setting up `useState`:

```tsx
  const [selectedRepos, setSelectedRepos] = useState<AddedRepositories>(() => {
    const sr = Object.values(values.repositories).reduce(
      (acc, repo) =>
        repo.orgName === orgData?.orgName ? { ...acc, [repo.id]: repo } : acc,
      {},
    );
    return sr;
  });
```

This way, it never runs when something changes in the "parent form" and doesn't require a rendering after this data is calculated.

debsmita1 commented on plugins/bulk-import/src/components/Repositories/DeleteRepositoryDialog.tsx:
```diff
@@ -70,13 +63,30 @@ const DeleteRepositoryDialog = ({
+    const value = await bulkImportApi.importByRepo(
+      repository.repoUrl || '',
+      repository?.defaultBranch || 'main',
+      'DELETE',
```
Changed it to `importActionOnRepo`

debsmita1 commented on plugins/bulk-import/src/components/Repositories/DeleteRepositoryDialog.tsx:
```diff
@@ -70,13 +63,30 @@ const DeleteRepositoryDialog = ({
+    setIsSubmitting(true);
+    const value = await bulkImportApi.importByRepo(
+      repository.repoUrl || '',
+      repository?.defaultBranch || 'main',
```
Handled the error

debsmita1 commented on plugins/bulk-import/src/hooks/useRepositories.ts:
```diff
@@ -0,0 +1,120 @@
+    totalRepositories: number;
+    totalOrganizations: number;
+  } | null;
+  error: Error | string[] | undefined;
```
Created an error type for this

debsmita1 commented on plugins/bulk-import/src/hooks/useRepositories.ts:
```diff
@@ -0,0 +1,120 @@
+      totalRepositories:
+        ((value as OrgAndRepoResponse)?.repositories || []).length > 0
+          ? (value as OrgAndRepoResponse)?.totalCount || 0
+          : 0,
```
I tried with `useAsync<() => Promise<OrgAndRepoResponse>>` and still got the type error. the `dataFetcher` returns data of 2 different types depending upon the `options` argument

jerolimov replied:
Got it. Yes, the issue is that `dataFetcher` returns different types depending on the input fields.

We don't need to address that know, but I like to pick this up as an example or experiment if we could use some API code generator.

jerolimov replied:
mark it as resolved to merge this PR, @debsmita1 will check if we can address this in a follow up PR.

debsmita1 commented on plugins/bulk-import/src/types/types.ts:
```diff
@@ -0,0 +1,107 @@
+
+import { Repository as RepositoryResponse } from './response-types';
+
+export type RepositoriesData = {
```
both

debsmita1 commented:
> I first take a look from the user perspective before I will review the code as well.
> 
> Just some small ideas/feedback, that we can also fix in a follow-up -- **_if_** @ShiranHi agrees!
> 
> 1. README. I was required to add this into my Root.tsx:
>    ```
>    <SidebarItem ... />
>    <SidebarItem ... />
>    <BulkImportIcon />
>    ```
>    
Done, changed it to `BulkImportSidebarItem`     
>          
>        
>    
>        
>      
>    Currently the `BulkImportIcon` is actual a SidebarItem. Please export just the icon as `BulkImportIcon` and if you want simplify the SidebarItem please export a `BulkImportSidebarItem`.



> 3. **UNSURE IF THIS IS STILL THE CASE, CAN'T OPEN THAT PAGE**
>    When editing a Bulk Import this link is also shown and should have also such an icon.
>    ![image](https://private-user-images.githubusercontent.com/139310/358211647-4a941341-c971-4a03-a10c-045086cce655.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNDY3MDUsIm5iZiI6MTcyNDA0NjQwNSwicGF0aCI6Ii8xMzkzMTAvMzU4MjExNjQ3LTRhOTQxMzQxLWM5NzEtNGEwMy1hMTBjLTA0NTA4NmNjZTY1NS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxOVQwNTQ2NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1iYmViMDI3MWExMWM5ZTNlYzQ4ODZhYzFkYzNmNDU3NzJjMTBiMDAxZmU2ZDgyYjcwYjFiYmNiZjdhN2UxNTVkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.76KGT5w_AFFWO9MLGPRxmpQoFS7KfQIb05c5x8GTli8)

Fixed it.
![Screenshot 2024-08-19 at 11 30 11 AM](https://github.com/user-attachments/assets/6eff4f56-3f11-43b2-bf8a-a2f1f97d55e8)




> 4. On the "Add repositories" the search field uses "Search" and placeholder, while the catalog and import list uses "Filter".
>    ![image](https://private-user-images.githubusercontent.com/139310/358211720-842b1e9c-9314-4e9d-99dd-2ee509bc05c6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNDY3MDUsIm5iZiI6MTcyNDA0NjQwNSwicGF0aCI6Ii8xMzkzMTAvMzU4MjExNzIwLTg0MmIxZTljLTkzMTQtNGU5ZC05OWRkLTJlZTUwOWJjMDVjNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxOVQwNTQ2NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0yMGYyMDVkYzUyN2FhZjJkZDY2YTRmZTIxMDRiMjcxZGJlNTA3OGZjNDkyMzBiOGQwZjc2ZjY1YmE1ZTNhNDMyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9._GYYD12zkvvEWMVHLdIjFcwSg6orcIiak93hyAe-NBk)
>    vs
>    ![image](https://private-user-images.githubusercontent.com/139310/358211787-76af575c-fcf3-4e9e-bb41-e361e39dccc6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNDY3MDUsIm5iZiI6MTcyNDA0NjQwNSwicGF0aCI6Ii8xMzkzMTAvMzU4MjExNzg3LTc2YWY1NzVjLWZjZjMtNGU5ZS1iYjQxLWUzNjFlMzlkY2NjNi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxOVQwNTQ2NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT05ZGUwOTNjNmVlZjRjZTU4NDgxOGM5MzBmMTYwNjMyODNlNTQ4MWJkZmFkNDZhODk2ZTgxZWMzM2Y5NjhiZGQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.V_WG6j1jxdyJZhOoUa_2c0eyI9QM23a5WrXYkZz_wSM)
>    Later I noticed that the search only works on the current page, which is IMO a reason to hide that filter at all. See below.

On the Repositories list page, I'm using Backstage's Table component, which doesn't allow props to modify the placeholder name. On the Add Repositories page, I'm using MUI table components to create a custom table header.


> 6. Didn't the first version only supports GitHub PRs? Should we rename Git to GitHub and disable the ServiceNow option for now?
>    ![image](https://private-user-images.githubusercontent.com/139310/358212517-0a9449e1-f56f-4ec7-91f3-e94f0c4dff5b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQwNDY3MDUsIm5iZiI6MTcyNDA0NjQwNSwicGF0aCI6Ii8xMzkzMTAvMzU4MjEyNTE3LTBhOTQ0OWUxLWY1NmYtNGVjNy05MWYzLWU5NGYwYzRkZmY1Yi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODE5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgxOVQwNTQ2NDVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zMmNlYWYxNGQzZTJmMzUyY2E1NzA3OWQ5M2EyZDA1YTJkYjk2MzA5MTczNzVlODFlOGQwY2FjYzI0MTEwNGZlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.CFnOoCLiJaCPMLKlr5XhGzyGnKoFm5e70PC9Xda651M)


Could you check if you are using the latest bulk import version ? 
the bulk import plugin was initially private to avoid version updates till the APIs were integrated but during the backstage upgrade it was changed to public.

debsmita1 commented on plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx:
```diff
@@ -404,101 +362,65 @@ export const RepositoriesTable = ({
+        ) : (
+          <Table
+            sx={{ minWidth: 750 }}
+            aria-labelledby={ariaLabel()}
```
removed this, we have `data-testId` here

jerolimov replied:
resolved

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesDrawer.tsx:
```diff
@@ -106,21 +115,21 @@ export const AddRepositoriesDrawer = ({
-              onClick={() => handleSelectRepoFromDrawer(selectedReposID)}
+              onClick={() => handleSelectRepoFromDrawer(selectedRepos)}
               className={classes.createButton}
               aria-labelledby="select-from-drawer"
```
resolved

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesDrawer.tsx:
```diff
@@ -129,7 +138,7 @@ export const AddRepositoriesDrawer = ({
           <span>
             <Button
-              aria-labelledby="cancel-drawer-select"
+              aria-labelledby="close-drawer"
```
resolved

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesFormFooter.tsx:
```diff
@@ -38,33 +43,39 @@ const useStyles = makeStyles(theme => ({
 }));
 
+const sPad = (repositories: AddedRepositories) =>
+  (Object.keys(repositories || [])?.length || 0) > 1 ? 's' : '';
```
resolved

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/AddRepositoriesFormFooter.tsx:
```diff
@@ -38,33 +43,39 @@ const useStyles = makeStyles(theme => ({
+    if (disableCreate) {
+      return 'Please wait until the catalog-info.yaml files are generated';
+    }
+    return '';
```
I'm sorry, but your new code will not work because you're checking toolTipTitle, which is an always-defined function.

Please check for `disableCreate` in the return statement. Remove the `toolTipTitle` function and set the title directly to the string that this function returns.

jerolimov replied:
Updated code should work

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/CatalogInfoStatus.tsx:
```diff
@@ -1,45 +1,95 @@
+  const { values, setFieldValue } =
+    useFormikContext<AddRepositoriesFormValues>();
+
+  React.useEffect(() => {
+    if (shouldExcludeRepositories(importStatus || '')) {
+      setFieldValue(`excludedRepositories.${data.id}`, {
+        status: importStatus,
+      });
+    }
+  }, [data.id, importStatus, setFieldValue, data.repoName]);
```
Ah, rechecked this after you explained that this is a component per row.

But a row should not change a formik state when it's rendered.

If the row shouldn't be selected on "Select all", then we should fix that callback on the table.

jerolimov commented on plugins/bulk-import/src/components/AddRepositories/RepositoriesTable.tsx:
```diff
@@ -69,83 +40,83 @@ export const RepositoriesTable = ({
-  useEffect(() => {
-    if (drawerOrganization) {
-      setTableData(orgReposData || []);
+  React.useEffect(() => {
+    if (showOrganizations) {
+      setOrgsData(prev => {
```
As we discussed, we should try to use formik in the drawer as well, maybe with a 2nd formik context.

But I'm accepting that we don't use formik there at the moment.

jerolimov commented on plugins/bulk-import/src/components/PreviewFile/PreviewPullRequest.tsx:
```diff
@@ -307,93 +352,102 @@ export const PreviewPullRequest = ({
+              delete pr[repoName]?.yaml?.spec?.owner;
+
+              setPullRequest(pr);
+              if (event.target.checked) {
+                const err = { ...formErrors };
+                delete err[repoName]?.entityOwner;
+                setFormErrors(err);
+              }
```
As we discussed, we should try to use formik in the drawer as well, maybe with a 2nd formik context.

But I'm accepting that we don't use formik there at the moment.

rm3l reviewed (commented):
@debsmita1 Could you rebase your PR onto `main`? I feel like the new dry-run checks I added in https://github.com/janus-idp/backstage-plugins/pull/1979 might cause an issue with PR creation (especially the one that checks the presence of a CODEOWNERS file).
I rebased locally to check and now PRs are not getting created: `CODEOWNERS_FILE_NOT_FOUND_IN_REPO` (added as a new check as we discussed). 

![image](https://github.com/user-attachments/assets/cd0d8adc-bc08-4bf7-beda-33137d34d3f7)

I think the frontend should enforce this only if the "Use CODEOWNERS file as Entity Owner" checkbox is selected. WDYT?

rm3l commented:
> @debsmita1 Could you rebase your PR onto `main`? I feel like the new dry-run checks I added in #1979 might cause an issue with PR creation (especially the one that checks the presence of a CODEOWNERS file). I rebased locally to check and now PRs are not getting created: `CODEOWNERS_FILE_NOT_FOUND_IN_REPO` (added as a new check as we discussed).
> 
> ![image](https://private-user-images.githubusercontent.com/593208/359459424-cd0d8adc-bc08-4bf7-beda-33137d34d3f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQxNjg5NjksIm5iZiI6MTcyNDE2ODY2OSwicGF0aCI6Ii81OTMyMDgvMzU5NDU5NDI0LWNkMGQ4YWRjLWJjMDgtNGJmNy1iZWRhLTMzMTM3ZDM0ZDNmNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIwJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMFQxNTQ0MjlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02MzM0ZWVmNzZjMjBiNTZkYTAzYzdlNDRmZWE4N2RiMTFkYTYzZmQ2YzNhNTMyODk4ZDYwYmU2NTA1MjI4MGQ3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.49guf9RkeXV_TgLM-Pwp5xQWDV2gcwlcbzWap-tAAPo)
> 
> I think the frontend should enforce this only if the "Use CODEOWNERS file as Entity Owner" checkbox is selected. WDYT?

As discussed, I've created https://github.com/janus-idp/backstage-plugins/pull/2068 so that clients can skip this new CODEOWNERS dry-run check as needed.

rm3l commented:
> I merged one of the PRs raised by bulk-import plugin, and the row disappeared from the "Added repositories" table. Is this expected? Also the one with "Finished and Ingested" from last test also disappeared. Both of the disappeared repositories displayed in the "Add repositories" page as unchecked which looks a bit weird.
> 

I think this potential issue of disappearing repos should be fixed with the changes in https://github.com/janus-idp/backstage-plugins/pull/2075
In its current implementation of the "listing all imports" endpoint, the backend starts by fetching repos, which might return a partial view depending on the number of repos and page and size requested.
https://github.com/janus-idp/backstage-plugins/pull/2075 should fix this, and drastically reduce the response time of this endpoint as well.

rm3l commented:
> @debsmita1 Could you rebase your PR onto `main`? I feel like the new dry-run checks I added in #1979 might cause an issue with PR creation (especially the one that checks the presence of a CODEOWNERS file). I rebased locally to check and now PRs are not getting created: `CODEOWNERS_FILE_NOT_FOUND_IN_REPO` (added as a new check as we discussed).
> 
> ![image](https://private-user-images.githubusercontent.com/593208/359459424-cd0d8adc-bc08-4bf7-beda-33137d34d3f7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQyNTU3MjUsIm5iZiI6MTcyNDI1NTQyNSwicGF0aCI6Ii81OTMyMDgvMzU5NDU5NDI0LWNkMGQ4YWRjLWJjMDgtNGJmNy1iZWRhLTMzMTM3ZDM0ZDNmNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwODIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDgyMVQxNTUwMjVaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mZWFlZjk3NDQwYTIwM2UyMmMwNGE1N2U3ZTI3ZTY2OTYyYWZkMzU1ZDIzMTMxMDg3ZGU4MjI3Zjk1NzczM2FiJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.tZXzLUgH7zvCTslhQ36jRVDcxVtkdaKI9OBLPyCeArg)
> 
> I think the frontend should enforce this only if the "Use CODEOWNERS file as Entity Owner" checkbox is selected. WDYT?

I just tested the updated branch, and import PR creation now works correctly with the conditional CODEOWNERS check. Thanks.

ciiay reviewed (commented):
Hi @debsmita1 , it works very good now. Awesome job 👏 
Just two concerns. 
- For users with permission of an added repo, will the "Edit" action being handled in another story?
- Delete action now with no processing indicator, which looks a bit unnatural. Maybe we can display the loading component for the added repo list?

debsmita1 commented:
> Hi @debsmita1 , it works very good now. Awesome job 👏 Just two concerns.
> 
> * For users with permission of an added repo, will the "Edit" action be handled in another story?

Yes, it is handled here https://github.com/janus-idp/backstage-plugins/pull/2034 and it is ready for review

> * Delete action now with no processing indicator, which looks a bit unnatural. Maybe we can display the loading component for the added repo list?

I can to see the processing indicator. Could you check if you are using the latest plugin version ?

<img width="732" alt="Screenshot 2024-08-22 at 12 02 41 PM" src="https://github.com/user-attachments/assets/86759bdc-5376-47ca-893f-8bee7642a3ba">

ciiay commented:
I checked out to your PR branch and I confirmed my local has your PR change to add the progressing icon in `DeleteRepositoryDialog.tsx`
```
          startIcon={
            isSubmitting && <CircularProgress size="20px" color="inherit" />
          }
```
Not sure why it doesn't show the progress icon for me

https://github.com/user-attachments/assets/d4b2cdae-d6f2-4254-b3f3-834e03485e83

All other changes are working great.

jerolimov commented on plugins/bulk-import/src/components/Repositories/RepositoriesListColumns.tsx:
```diff
@@ -41,19 +53,24 @@ export const columns: TableColumn<AddRepositoriesData>[] = [
     title: 'Last updated',
-    field: 'lastUpdated',
+    field: 'catalogInfoYaml.lastUpdated',
     type: 'string',
```
Show a formatted date:

```suggestion
    type: 'datetime',
```

jerolimov replied:
mark it as resolved to merge this PR, @debsmita1 will check if we can address this in a follow up PR.

jerolimov commented:
@debsmita1 @rm3l @ShiranHi 

### Open comments

From https://github.com/janus-idp/backstage-plugins/pull/2013#pullrequestreview-2240296261

* [ ] https://github.com/janus-idp/backstage-plugins/pull/2013#discussion_r1718356875
* [ ] https://github.com/janus-idp/backstage-plugins/pull/2013#discussion_r1718378952
* [ ] https://github.com/janus-idp/backstage-plugins/pull/2013#discussion_r1718380839
* [ ] https://github.com/janus-idp/backstage-plugins/pull/2013#discussion_r1718487019

Latest comment https://github.com/janus-idp/backstage-plugins/pull/2013#pullrequestreview-2260014464

* [ ] https://github.com/janus-idp/backstage-plugins/pull/2013#discussion_r1730831753

### Confusing or missing information from a user perspective:

* imo: Search filter filters only the name, but it would be great if it filters the organization or complete URL as well. /cc @ShiranHi 
* No status filter (for the catalog-info.yaml column) --- I know that's complicated with the current architecture, we should address this now. But it would be nice to have this in the backlog, or?
* `Waiting for PR Approval` status doesn't provide a link to open that PR. /cc @ShiranHi 
  * On the import "Added repositories" page it shows an action to open the PR. But it says "catalog-info.yaml". It would be nice to link the PR directly from the Status cell.
  * In the import a new repository flow there is no option to open that PR.
* Action "View catalog-info.yaml file" does nothing if the PR is merged (Job status "Finished and Ingested")
* Latest update(d) column doesn't use a formatted date, see comment above
* Delete repository mean "Remove catalog entry and delete PR?" Esp. the current error message "Unable to remove repository. Repository default branch is missing" made me worried for a second that you try to remove my repository. ;)
* When I **close a PR** (which means to me I don't want to merge the catalog-info.yaml) /cc @debsmita1 @rm3l 
  * the status shows nothing
  * the Action "View catalog-info.yaml file" does nothing on click

### Other ideas

* Using react-query might reduce the number of API calls needed to fetch the same information. For example, when the user switches between repositories and organizations or when using the filter with the maximum page size, the rendering triggers the same API calls repeatedly.

debsmita1 commented:
> * `Waiting for PR Approval` status doesn't provide a link to open that PR. /cc @ShiranHi
>   
>   * On the import "Added repositories" page it shows an action to open the PR. But it says "catalog-info.yaml". It would be nice to link the PR directly from the Status cell.
>   * In the import of a new repository flow, there is no option to open that PR.
> * Action "View catalog-info.yaml file" does nothing if the PR is merged (Job status "Finished and Ingested")
> * Latest update(d) column doesn't use a formatted date, see comment above
> * Delete repository means "Remove catalog entry and delete PR?" Esp. the current error message "Unable to remove repository. Repository default branch is missing" made me worried for a second that you try to remove my repository. ;)
> * When I **close a PR** (which means to me I don't want to merge the catalog-info.yaml) /cc @debsmita1 @rm3l
>   
>   * The status shows nothing
>   * the Action "View catalog-info.yaml file" does nothing on click

I have handled the above in this PR https://github.com/janus-idp/backstage-plugins/pull/2034

jerolimov reviewed (approved):
As discussed on slack, this PR is fine to get in as you already have opened a follow up PR #2034

/lgtm

jerolimov commented:
/unassign

ShiranHi commented:
> `Waiting for PR Approval` status doesn't provide a link to open that PR. /cc @ShiranHi
> * On the import "Added repositories" page it shows an action to open the PR. But it says "catalog-info.yaml". It would be nice to link the PR directly from the Status cell.
> * In the import of a new repository flow, there is no option to open that PR.

This flow focuses on adding new repositories and users can link the URL if they want to navigate to that specific repo. However, I prefer to keep users within the form editing process during this "Add" flow. If the repository has been added previously, users can manage it from the main table, where they can view all the added repositories. 
 
>Delete repository means "Remove catalog entry and delete PR?" Esp. the current error message "Unable to remove repository. Repository default branch is missing" made me worried for a second that you try to remove my repository. ;)

Removing a repository means it will no longer be visible on the Catalog page ([Figma](https://www.figma.com/design/09Iltd57RiuqOtKWkjy8MH/Bulk-import?node-id=473-668&t=XtbQwzyEahuzkTpD-1)).

>A status filter (for the catalog-info.yaml column) would be nice. In the best case, that shows me 10 repos added, 30 repos have pending PRs, and 1000 repos haven't started the import yet. I know that's complicated with the current architecture; we should address this now. But it would be nice to have this in the backlog, or?

+1 :)
